opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt

```diff
@@ -0,0 +1,571 @@
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+package de.geeksfactory.opacclient.apis
+
+import de.geeksfactory.opacclient.i18n.StringProvider
+import de.geeksfactory.opacclient.networking.HttpClientFactory
+import de.geeksfactory.opacclient.objects.*
+import de.geeksfactory.opacclient.searchfields.DropdownSearchField
+import de.geeksfactory.opacclient.searchfields.SearchField
+import de.geeksfactory.opacclient.searchfields.SearchQuery
+import de.geeksfactory.opacclient.searchfields.TextSearchField
+import de.geeksfactory.opacclient.utils.get
+import de.geeksfactory.opacclient.utils.html
+import de.geeksfactory.opacclient.utils.text
+import okhttp3.FormBody
+import okhttp3.HttpUrl
+import org.joda.time.LocalDate
+import org.joda.time.format.DateTimeFormat
+import org.joda.time.format.DateTimeFormatter
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import org.jsoup.Jsoup
+import org.jsoup.parser.Parser
+
+/**
+ * OpacApi implementation for SLUB. https://slub-dresden.de
+ *
+ * @author Steffen Rehberg, Jan 2019
+ */
+open class SLUB : OkHttpBaseApi() {
+    protected lateinit var baseurl: String
+    protected lateinit var illRenewUrl: String
+    private val ENCODING = "UTF-8"
+    protected lateinit var query: List<SearchQuery>
+
+    private val mediaTypes = mapOf(
+            "Article, E-Article" to SearchResult.MediaType.EDOC,
+            "Book, E-Book" to SearchResult.MediaType.BOOK,
+            "Video" to SearchResult.MediaType.EVIDEO,
+            "Thesis" to SearchResult.MediaType.BOOK,
+            "Manuscript" to SearchResult.MediaType.BOOK,
+            "Musical Score" to SearchResult.MediaType.SCORE_MUSIC,
+            "Website" to SearchResult.MediaType.URL,
+            "Journal, E-Journal" to SearchResult.MediaType.NEWSPAPER,
+            "Map" to SearchResult.MediaType.MAP,
+            "Audio" to SearchResult.MediaType.EAUDIO,
+            "Electronic Resource (Data Carrier)" to SearchResult.MediaType.EAUDIO,
+            "Image" to SearchResult.MediaType.ART,
+            "Microform" to SearchResult.MediaType.MICROFORM,
+            "Visual Media" to SearchResult.MediaType.ART
+    )
+
+    private val fieldCaptions = mapOf(
+            "format" to "Medientyp",
+            "title" to "Titel",
+            "contributor" to "Beteiligte",
+            "publisher" to "Erschienen",
+            "ispartof" to "Erschienen in",
+            "identifier" to "ISBN",
+            "language" to "Sprache",
+            "subject" to "Schlagwörter",
+            "description" to "Beschreibung",
+            "linksRelated" to "Info",
+            "linksAccess" to "Zugang",
+            "linksGeneral" to "Link"
+    )
+
+    private val ACTION_COPY = OpacApi.MultiStepResult.ACTION_USER + 1
+    private val pickupPoints = mapOf(
```

Comment:
I'm afraid the clear texts for the pickup point codes will have to be hardcoded here. The API only sends the codes (without the text). The web catalog does of course show the texts when you're going to reserve an item or put a hold on it but it only shows the points available for this particular item (as a rule, just 1 or 2 possible pickup points). So I had to compile this list by trying to reserve a lot of different items and then see what code - clear texts combinations the web site offered.  
New pickup points, which are not in the map, will get their code as clear text:
https://github.com/opacapp/opacclient/blob/08dfa091c944603cda0a76cd025fb2c200d7ba14/opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/SLUB.kt#L334
When this happens this new code must be looked up in the web catalog by trying to make the reservation and then update the code here.  
I understand that this is not ideal but it's the price for using the super fast API rather than scraping the relatively slow web site.

Reply:
Okay, in that case I think it's fine. Thanks for the explanation!